File: Cap 12/Topico 12.7/Polimorfismo/Universal/Parametrico/polimunipar.js
/*
 * Livro .....: Algoritmos
 *              Lógica para Desenvolvimento de Programação de
 *              Computadores
 * Autores ...: José Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Érica
 *
 * Capítulo ..: 12 - Tópico 12.7
 * 
 * Programa ..: POLIFORMISMO UNIVERSAL PARAMÉTRICO
 * 
 */

class CLS_PESSOA {
  constructor(NOME) {
    this.NOME = NOME;
  }

  PEGNOME() {
    console.log(this.NOME);
  }

  PROFISSAO() {
    console.log("Atividade profissional desconhecida.");
  }
}

class CLS_MEDICO extends CLS_PESSOA {
  constructor(NOME) {
    super(NOME);
  }

  PROFISSAO() {
    console.log("Você possui formação em medicina.");
  }
}

class CLS_ADVOGADO extends CLS_PESSOA {
  constructor(NOME) {
    super(NOME);
  }

  PROFISSAO() {
    console.log("Você possui formação em advocacia.");
  }
}

let PROFISSIONAL;
let PROFIS1 = new CLS_ADVOGADO("Carlos Alberto");
let PROFIS2 = new CLS_MEDICO("Paulo Malaquias");
let PROFIS3 = new CLS_PESSOA("Martha Nepomuceno");

PROFISSIONAL = PROFIS1;
PROFISSIONAL.PEGNOME();
PROFISSIONAL.PROFISSAO();
console.log();
PROFISSIONAL = PROFIS2;
PROFISSIONAL.PEGNOME();
PROFISSIONAL.PROFISSAO();
console.log();
PROFISSIONAL = PROFIS3;
PROFISSIONAL.PEGNOME();
PROFISSIONAL.PROFISSAO();

File: Cap 12/Topico 12.7/Polimorfismo/Universal/Parametrico/polimunipar.pl
=pod

 * Livro .....: Algoritmos
 *              Lógica para Desenvolvimento de Programação de
 *              Computadores
 * Autores ...: José Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Érica
 *
 * Capítulo ..: 12 - Tópico 12.7
 * 
 * Programa ..: POLIFORMISMO UNIVERSAL PARAMÉTRICO

=cut

use strict;
use warnings;
use feature 'say';

package CLS_PESSOA {
    sub new {
        my ($class, $nome) = @_;
        return bless {
            NOME => $nome
        }, $class;
    }

    sub PEGNOME {
        my $self = shift;
        say $self->{NOME};
    }

    sub PROFISSAO {
        say "Atividade profissional desconhecida.";
    }
}

package CLS_MEDICO {
    use parent -norequire, 'CLS_PESSOA';

    sub new {
        my ($class, $nome) = @_;
        return bless $class->SUPER::new($nome), $class;
    }

    sub PROFISSAO {
        say "Você possui formação em medicina.";
    }
}

package CLS_ADVOGADO {
    use parent -norequire, 'CLS_PESSOA';

    sub new {
        my ($class, $nome) = @_;
        return bless $class->SUPER::new($nome), $class;
    }

    sub PROFISSAO {
        say "Você possui formação em advocacia.";
    }
}

my $PROFISSIONAL;
my $PROFIS1 = CLS_ADVOGADO->new("Carlos Alberto");
my $PROFIS2 = CLS_MEDICO->new("Paulo Malaquias");
my $PROFIS3 = CLS_PESSOA->new("Martha Nepomuceno");

$PROFISSIONAL = $PROFIS1;
$PROFISSIONAL->PEGNOME();
$PROFISSIONAL->PROFISSAO();
say '';

$PROFISSIONAL = $PROFIS2;
$PROFISSIONAL->PEGNOME();
$PROFISSIONAL->PROFISSAO();
say '';

$PROFISSIONAL = $PROFIS3;
$PROFISSIONAL->PEGNOME();
$PROFISSIONAL->PROFISSAO();
